sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,543 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+#define OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+
+#include <stdint.h>
+
+/**
+ * RISC-V Enhanced Physical Memory Protection (EPMP).
+ *
+ * Specifications:
+ *  - PMP Enhancements for memory access and execution prevention on Machine
+ *    mode
+ *    https://docs.google.com/document/d/1Mh_aiHYxemL0umN3GTTw8vsbmzHZ_nxZXgjgOUzbvc8
+ *  - RISC-V Privileged Specfication (v. 20190608)
+ *    https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMFDQC-and-Priv-v1.11/riscv-privileged-20190608.pdf
+ *  - Ibex PMP
+ *    https://ibex-core.readthedocs.io/en/latest/03_reference/pmp.html
+ *
+ * Assumptions (should be initialized in assembly but can be verified using
+ * `epmp_check`):
+ *   * Machine Mode Whitelist Policy is enabled (mseccfg.MMWP = 1)
+ *   * Machine Mode Lockdown is disabled (mseccfg.MML = 0)
+ */
+
+/**
+ * EPMP constants.
+ *
+ * TODO: there will probably be more of these.
+ */
+typedef enum {
+  kEpmpGranularity = 0,
+  kEpmpNumRegions = 16,
+} epmp_constants_t;
+
+/**
+ * EPMP entry permissions.
+ *
+ * Entries configured with locked permissions may only be modified
+ * if Rule Locking Bypass is set (mseccfg.RLB = 1).
+ *
+ * When Machine Mode Lockdown is disabled (mseccfg.MML = 0) the
+ * combination R=0 W=1 is reserved. This is the assumed state and so
+ * it is not possible to set these values. The combination R=1 W=1
+ * X=1 has also been reserved by this library and may not be used
+ * (TODO: double check this is OK).
+ *
+ * An entry may only be configured with unlocked permissions if the
+ * entry is also configured as OFF.
+ *
+ * TODO: read-write-execute has been deliberately left out too since
+ * it shouldn't be needed. Is that OK?
+ *
+ * Note: these permissions have different meanings when Machine Mode
+ * Lockdown (mseccfg.MML) is set.
+ */
+typedef enum {
+  kEpmpPermUnlocked,
```

Comment:
The intent is that `Unlocked` is the 0 value for LRWX (i.e. unlocked, no permissions set). I'll add a comment in the next round. We could make it `UnlockedNoAccess` but since M-mode will have full access that might be confusing.

Reply:
>  since M-mode will have full access that might be confusing

Is this true when `MMWP = 1`?

Reply:
My understanding is that `MMWP = 1` only has an effect if there are no PMP entries matching a given region. When `MML = 0` a non-OFF unlocked entry will match but it won't restrict M-mode in any way and so is equivalent to full read-write-execute access.

@GregAC is my understanding correct? The specifications seem ambiguous on this one. Maybe I am missing something.

Reply:
Yes I believe so, key line from the spec (so behavior with `MML = 0`):

> If a PMP entry matches all bytes of an access, then the L, R, W, and X bits determine whether the
access succeeds or fails. If the L bit is clear and the privilege mode of the access is M, the access
succeeds.

Default deny when `MMWP = 1` only applies if there's nothing allowing the access and in this case there would be an entry allowing the access.

Reply:
IMHO, it would be nice to have more functionally descriptive names for the enums (not sure if this is a good idea or would be too confusing). Should we consider defining two constants with more descriptive names (for `MML` `0` and `1`) with the same value. We could also have two different enum types and function that translates them to a common type like this one.

Reply:
I did originally have more descriptive names but they did end up being very verbose. I think we can rename these later if we think the current names are too light on information, that shouldn't be too onerous.

Reply:
> Should we consider defining two constants with more descriptive names (for MML 0 and 1) with the same value.

We may want to add this later but let's see how far we get without it.